pkg/agent/puller.go

```diff
@@ -1,73 +1,130 @@
-		return fmt.Errorf("failing to delete model directory: %v", err)
+}
+
+type ModelChannel struct {
+	modelOps    chan *v1.ModelSpec
+	opsInFlight int
+}
+
+func (p *Puller) enqueueModelOp(modelName string, modelSpec *v1.ModelSpec) {
+	modelChan, ok := p.channelMap[modelName]
+	if !ok {
+		modelChan = ModelChannel{
+			modelOps: make(chan *v1.ModelSpec, 8),
+		}
+		go p.modelProcessor(modelName, modelChan.modelOps)
+		p.channelMap[modelName] = modelChan
+	}
+	modelChan.opsInFlight += 1
+	modelChan.modelOps <- modelSpec
+}
+
+func (p *Puller) modelOpComplete(modelName string, closed bool) {
+	modelChan, ok := p.channelMap[modelName]
+	if ok {
+		modelChan.opsInFlight -= 1
+		if modelChan.opsInFlight == 0 {
+			close(modelChan.modelOps)
+			delete(p.channelMap, modelName)
+			if closed && len(p.channelMap) == 0 {
+				// this was the final completion, close the channel
+				close(p.completions)
+			}
+		}
+	} else {
+		//TODO log warning, shouldn't happen
 	}
-	return nil
 }
 
-func (p *Puller) modelProcessor(modelName string, events chan EventWrapper) {
+func (p *Puller) modelProcessor(modelName string, ops <-chan *v1.ModelSpec) {
 	log.Println("worker for", modelName, "is initialized")
 	// TODO: Instead of going through each event, one-by-one, we need to drain and combine
 	// this is important for handling Load --> Unload requests sent in tandem
 	// Load --> Unload = 0 (cancel first load)
 	// Load --> Unload --> Load = 1 Load (cancel second load?)
-	for event := range events {
-		log.Println("worker", modelName, "started  job", event)
-		switch event.LoadState {
-		case ShouldLoad:
-			log.Println("Should download", event.ModelSpec.StorageURI)
-			err := p.Downloader.DownloadModel(event)
+	for modelSpec := range ops {
+		if modelSpec != &remove {
+			// Load
+			log.Println("Should download", modelSpec.StorageURI)
+			err := p.Downloader.DownloadModel(modelName, modelSpec)
 			if err != nil {
-				log.Println("worker failed on", event, "because: ", err)
+				log.Println("Download of model", modelName, "failed because: ", err)
 			} else {
 				// If there is an error, we will NOT send a request. As such, to know about errors, you will
 				// need to call the error endpoint of the puller
 				// TODO: Do request logic
-				log.Println("Now doing a request on", event)
+				log.Println("Now doing load request for", modelName)
 			}
-		case ShouldUnload:
+		} else {
+			// Unload
 			// TODO: Do request logic
-			log.Println("Now doing a request on", event)
+			log.Println("Now doing unload request for", modelName)
 			// If there is an error, we will NOT do a delete... that could be problematic
-			log.Println("Should unload", event.ModelName)
-			if err := p.RemoveModel(event.ModelName); err != nil {
-				log.Println("worker failed on", event, "because: ", err)
+			log.Println("Should unload", modelName)
+			if err := storage.RemoveDir(filepath.Join(p.Downloader.ModelDir, modelName)); err != nil {
```

Comment:
Should we not keep this logic in a method that will probably be expanded upon, or is the extent of the storage deleting a model, going to be deleting the `dir`

Reply:
Probably yes, maybe even as a function on the Downloader for symmetry. I can't envisage what else would be done for this apart from deleting the model's dir, but it will probably be important to ensure that the "success" file is deleted first before the rest of it.